sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/UdafImpl.java

```diff
@@ -58,11 +58,23 @@ public String getName() {
           }
 
           public RelDataType getType(RelDataTypeFactory typeFactory) {
-            //the first generic type of CombineFn is the input type.
-            ParameterizedType parameterizedType = (ParameterizedType) combineFn.getClass()
-                .getGenericSuperclass();
+            ParameterizedType parameterizedType = findCombineFnSuperClass();
             return typeFactory.createJavaType(
-                (Class) parameterizedType.getActualTypeArguments()[0]);
+              (Class) parameterizedType.getActualTypeArguments()[0]);
+          }
+
+          private ParameterizedType findCombineFnSuperClass() {
+            Class clazz = combineFn.getClass();
+            while (!clazz.getSuperclass().equals(CombineFn.class)) {
+              clazz = clazz.getSuperclass();
+            }
+
+            if (!(clazz.getGenericSuperclass() instanceof ParameterizedType)) {
+              throw new IllegalStateException(
```

Comment:
not sure this is the best exception to throw. Please provide guidance. Should there be a custom Exception class?

Reply:
I think `IllegalStateException` is fine here